transport/azure/azure.go

```diff
@@ -32,44 +34,65 @@ func New() (http.RoundTripper, error) {
 		return nil, err
 	}
 
-	serviceURL := azblob.NewServiceURL(*endpointURL, pipeline)
+	if config.ABSKey != "" {
+		sharedKeyCredential, err = azblob.NewSharedKeyCredential(config.ABSName, config.ABSKey)
+		if err != nil {
+			return nil, err
+		}
 
-	return transport{&serviceURL}, nil
-}
+		client, err = azblob.NewClientWithSharedKeyCredential(endpointURL.String(), sharedKeyCredential, nil)
+	} else {
+		defaultAzureCredential, err = azidentity.NewDefaultAzureCredential(nil)
+		if err != nil {
+			return nil, err
+		}
 
-func (t transport) RoundTrip(req *http.Request) (resp *http.Response, err error) {
-	containerURL := t.serviceURL.NewContainerURL(strings.ToLower(req.URL.Host))
-	blobURL := containerURL.NewBlockBlobURL(strings.TrimPrefix(req.URL.Path, "/"))
+		client, err = azblob.NewClient(endpointURL.String(), defaultAzureCredential, nil)
+	}
 
-	start, end, err := httprange.Parse(req.Header.Get("Range"))
 	if err != nil {
-		return httprange.InvalidHTTPRangeResponse(req), nil
+		return nil, err
 	}
 
-	length := end - start + 1
-	if end <= 0 {
-		length = azblob.CountToEnd
-	}
+	return transport{client}, nil
+}
 
-	get, err := blobURL.Download(req.Context(), start, length, azblob.BlobAccessConditions{}, false, azblob.ClientProvidedKeyOptions{})
+func (t transport) RoundTrip(req *http.Request) (resp *http.Response, err error) {
+	container := req.URL.Host
+	key := req.URL.Path
+
+	header := make(http.Header)
+
+	result, err := t.svc.DownloadStream(req.Context(), container, strings.TrimPrefix(key, "/"), nil)
 	if err != nil {
-		return nil, err
+		if azError, ok := err.(*azcore.ResponseError); !ok || azError.StatusCode < 100 || azError.StatusCode == 301 {
+			return nil, err
+		} else {
+			return &http.Response{
+				StatusCode:    azError.StatusCode,
+				Proto:         "HTTP/1.0",
+				ProtoMajor:    1,
+				ProtoMinor:    0,
+				Header:        header,
+				ContentLength: *result.ContentLength,
+				Body:          ctxreader.New(req.Context(), result.Body, true),
+				Close:         true,
+				Request:       req,
+			}, nil
+		}
 	}
 
-	if config.ETagEnabled && start == 0 && end == azblob.CountToEnd {
-		etag := string(get.ETag())
-
-		if etag == req.Header.Get("If-None-Match") {
-			if body := get.Response().Body; body != nil {
-				get.Response().Body.Close()
-			}
+	if config.ETagEnabled {
+		azETag := string(*result.ETag)
```

Comment:
Forgive my assumptions here as I'm relatively new to the language, but in what scenario would this actually happen? Unless I'm understanding something wrong it could never be `nil` because it is not a pointer so its default value should be false if not set.

Based on the usage of this particular variable in both `s3.go` and `gcs.go` there is only a simple conditional check to see if it is `true`.

Reply:
`result.ETag` is a pointer to `string`, so it can be `nil`. If you try to unref `nil`, you'll get SEGFAULT. Example: https://go.dev/play/.

Reply:
I misinterpreted your original comment, my apologies. I'll push up the refactor shortly.